### ISSUE TITLE
cannot load (certain?) unnotarized externals on macOS

### ISSUE THREAD

**Issue by Spacechild1** [pure-data/pure-data#2166]
Hi, I have received several reports from users that they cannot use `[vstplugin~]` on macOS without circumventing the GateKeeper.

<img width="372" alt="w_allow" src="https://github.com/pure-data/pure-data/assets/16126632/817290d6-696b-41f2-a416-334437a4cd37">

![pd_log_anon](https://github.com/pure-data/pure-data/assets/16126632/6bff34da-2f23-490a-82e0-bdd9c6a3d3de)

`[vstplugin~]` is signed with the IEM's Apple ID (see https://git.iem.at/pd/vstplugin/-/commit/43c877b516a9132754488dc08c4048317a01ce43), but not notarized. However, the Pd app should contain the necessary entitlements to load unnotarized plugins.

Strangely, on my Intel MacBook (macOS 13.1) I can open it just fine. All the bug reports I get are from Apple Silicon users, so that might be a hint. Does anyone have problems with unnotarized externals on Apple Silicon?

It would be great if Apple users (Intel or Apple Silicon) could test and report back. You can get `[vstplugin~]` binaries either from Deken or from the release page: https://git.iem.at/pd/vstplugin/-/releases.

**Comment by ben-wes** [comment]
i had seen this warning before here on silicon (m2) before (although i don't remember where i got the built from back then). it was possible to give permissions to these files through the security settings though.

now i tested with a new installation from deken and didn't see any warnings.

**Comment by umlaeute** [comment]
while *we* (that is: those providing the "Pure Data" application) cannot do anything about this, external creators (those providing a pre-compiled external for macOS) **can** do something about this: sign and notarize the binaries.

i'm just adding this information, so anybody who is distributing externals and is willing to pay for an Apple Developer ID knows what the "correct way (approved by Apple Inc.) is.

also note, that the problem only shows, if the external is itself trying to open additional libraries (explicitly via `dlopen()` or similar; afaict this is not a problem for externals that depend on 3rd party libraries that opened via standard dynamic runtime linking)


finally, for completeness sake: as mentioned in https://github.com/supercollider/supercollider/issues/6132#issuecomment-1880668463, the problem is most prominent when downloading externals via *Safari* (which quarantines un-notarized binaries). As of today, other browsers (like *Firefox*, and luckily *deken*) do not quarantine the binaries, so they stay usable.

**Comment by Spacechild1** [comment]
> also note, that the problem only shows, if the external is itself trying to open additional libraries

How did you come to this conclusion? This problem really concerns *all* externals. Just to double-check, I downloaded the latest [aoo] and [zexy] libraries directly from [deken.puredata.info](https://deken.puredata.info/), first with Safari, then with Firefox. The Safari downloads refuse to load and give security popups, the Firefox downloads work just fine. ([aoo] and [zexy], to my knowledge, do not try to load any libraries.)

**Comment by umlaeute** [comment]
> How did you come to this conclusion?

obviously i was wrong then, thanks for clarifying.

**Comment by Spacechild1** [comment]
> i'm just adding this information, so anybody who is distributing externals and is willing to pay for an Apple Developer ID knows what the "correct way (approved by Apple Inc.)" is.

Yes, I explicitly note this in https://github.com/supercollider/supercollider/issues/6132#issuecomment-1880668463. However, I want to stress again that I don't think this is a practical solution. I'd even consider it immoral to do so because it puts even more burden on other casual open source developers. IMO we should completely boycott plugin notarization and educate the users instead.

**Comment by umlaeute** [comment]
> Yes, I explicitly note this in ...

yes. i just wanted to summarize it *here* (hating to have to go through zillions of references pages to get the actual information).

regarding morality: i feel your pain.

however, i don't really agree on the practical consequences:
- imho, notarization is actually more useful than codesigning (it's practically a cloud-based malware scan; although run by some big-tech E-corp, who can decide at will what a "malware" is)
- the hurdle for notarization (the requirement for a payed-for apple developer ID) is practically the same as for codesigning (so it's not just *notarization* that is evil)
- it's obviously in E-corp's interest to require codesigned/notarized binaries. so i think it will only get worse (e.g. i don't see why they won't lock down the system even more, e.g. to require *all* binaries to be signed/notarized regardless of whether they were obtained from a download,a USB stick or conjured out of thin air; so educating the users is bound to go stale)
- i do not see how a simple refusal to play by E-corp's rules is going to improve the situation

E-corp is not forcing anybody to write software that runs on their systems. What is forcing us is that we do care for a group of users who spend money on shiny things and expect "casual open source developers" to spend *their* spare money for the users' wellbeing.

I don't have an answer.

**Comment by Spacechild1** [comment]
>  i don't see why they won't lock down the system even more

That might happen eventually, but until then we can still resist. IMO there is no need for preemptive obedience ("vorauseilender Gehorsam").

> i do not see how a simple refusal to play by E-corp's rules is going to improve the situation

Because it puts everyone in the same boat. Once some external authors (with institutional support) start to notarize their plugins, it puts additional pressure on other developers to do the same.

> What is forcing us is that we do care for a group of users who spend money on shiny things and expect "casual open source developers" to spend their spare money for the users' wellbeing.

I don't really blame the users, I think most are not even aware of these issues. That's why I think we need to raise awareness! In fact, even Apple power users could not explain to me what was going on regarding these security warnings. For 3 years I have been trying to understand what's going on these MacBooks and even Apple power users couldn't explain it to me. I only got the crucial hint yesterday from some random blog post. That's why I am so frustrated and angry.

**Comment by Spacechild1** [comment]
BTW, I found this very helpful page: https://disable-gatekeeper.github.io/ :)

I discovered this through https://www.octasine.com/.
